### DISCUSSION TITLE
"Das Versenden einer E-Mail ist fehlgeschlagen." - how to find the bug?

### DISCUSSION THREAD

**Discussion by navlisilvan** [pretix/pretix#2530] [Q&A]
Hi,
we found out that a E-mail to all customers was not send to minimum one customer.
Mailbox was not full, "older" mails reached the customer.
In the "Bestellverlauf"-list there is twice the notice "Das Versenden einer E-Mail ist fehlgeschlagen. " with the same time.
(Pretix 4.7.1, batch email plugin 1.01 installed but not used for this mail)

Is there a way to find out, why it was not possible to send this mail? And how to find out wich other customers also didn´t get this mail?

Thank you very much!

**Comment by raphaelm** [comment]
If you are in "admin mode", there should be a button with the log entry to inspect it for more details. You can additionally check pretix log for that time.

**Reply by navlisilvan** [reply]
ah, thx.

```
{
  "invoices": [],
  "message": "4.7.0 xx.kasserver.com Error: too many connections from xx.xx.xx.xx",
  "recipient": "",
  "subject": "SMTP code 421, max retries exceeded"
}
```

hm. Is there a way to configure how many connections Pretix is using? And Is there a way to see if other Mails were also bounced because of this?

**Reply by raphaelm** [reply]
No, unfortunately there is currently no configuration for that as it's hard to do in the general-case setup. Theoretically, you could split your `pretix-worker` process int separate workers for multiple queues and limit the concurrency of the mail queue. This is documented here:
https://docs.pretix.eu/en/latest/admin/scaling.html#pretix-worker

> And Is there a way to see if other Mails were also bounced because of this?

Not systematically, you could look at the log entries on event level.

**Reply by navlisilvan** [reply]
But why does this happen? It was luck to found this so i don´t know it it was already happening.
I just sent an email to all customers via Pretix, nothing more. The pretix-worker should not open more than three connections, right? Or what could I have done wrong?

**Reply by raphaelm** [reply]
Where do you get the idea that it should not open more than 3 connections?

**Reply by navlisilvan** [reply]
If All-Inkl limits the maximum connections, maybe other hosters will do the same and then many users would have the problem when Pretix opens more than three connections, or am i wrong?

**Reply by raphaelm** [reply]
How do you come up with the number 3?

**Reply by raphaelm** [reply]
The problem is that pretix can be installed across many different servers and can handle many events and smtp servers. Coordinating that there are never more than 3 connections to the same server would be quite complex. Generally never sending more thant 3 emails at the same would barely be an option on large-scale installations.

**Reply by navlisilvan** [reply]
I found the number 3 in a forum and All-inkl has just confirmed it to me that there is this limit with them:

"Unser Server lässt 3 gleichzeitige SMTP-Verbindungen pro Client und Postfach zu.
Es gibt noch weitere SMTP-Limits, welche Sie (falls möglich) auch in Ihrem Programm hinterlegen könnten:
 max. 1000 E-Mails in 10 Minuten (bei Versand über SMTP)
 max. 400 Verbindungen in 10 Minuten
 max. 100 MB - Größe einer Einzelmail"
 
We have about 500 customers in one event.

So the function "E-Mail verschicken an - Jeden, der ein Ticket bestellt hat" is not useable because pretix-worker starts to open many smtp-connections at the same time? Or does it open a new connection with every email so that the 400-limit is exceeded?

**Reply by raphaelm** [reply]
As I read your error message, I assume you've ran into the limit of three. However it's also likely to run into the limits of 400 and 1000 in a medium sized event.

Honestly, my opinion about this is: all-inkl apparently isn't suited for mass email. GMail and Hotmail have similar limitations. All of these are built for regular office mail, not transactional mail or mass email.

As you can also already see from what you quoted, every email provider seems to define these limits in different ways. Sometimes it's connections, sometimes email, with different sizes and time intervals. Sometimes per client, sometimes per mailbox, sometimes a combination.

Also, the limits are often poorely documented, as "I found it in a forum" shows.

Also, the error messages if the limits are exceeded are not standardized and cannot be parsed and handled automatically.

Building a feature into pretix that allows to respect all possible types of limits will be hard and complicated and I'm not sure we want to go down that route.
